---
title: "F2 adaptations to dominant markers"
date: "`r Sys.Date()`"
author: "[Statistical Genetics Lab](http://statgen.esalq.usp.br) <br/> Department of Genetics <br/> Luiz de Queiroz College of Agriculture <br/> University of SÃ£o Paulo"
output:
    rmdformats::readthedown:
      css: ../readthedownstatgen.css
---

```{r knitr_init, echo=FALSE, cache=FALSE}
library(knitr)
library(rmdformats)

## Global options
options(max.print="75")

opts_knit$set(width=75)
```

One big update needed to be done to OneMap proper deal with dominant markers in F2 population. The modification was apply phase estimation also to F2 populations using the same background algorithms used for outcrossing species. This modification was needed to estimate the correct phase between dominant and codominant markers and also to distinguish phase between the heterozygotes progeny in codominant markers. In this new version, we want to provide to users progeny haplotypes, and for that, it's important to distinguish phase also in F2 intercross. 

**Warning**: Because all the modifications, users can find differences in estimations for this type of populations between this and older versions of OneMap.

Here we perform some simulations to show this differences and the need for updates. Two scenarios will be simulated, one with only codominant markers and other with codominant and dominant markers. The maps are composed by three metacentric chromosomes with 100 cM and a population size of 200 individuals. And:

* Scenario 1: 150 codominant (A.H.B) markers, 50 per chromosome and no missing data
* Scenario 2: 50 codominant (A.H.B), 50 D.B markers and 50 C.A markers, 50 markers per chromosome and no missing data

Both scenarios will be evaluated by old and new approach.

# Packages

```{r}
library(onemap)
```

# Scenario 1

## Simulation

```{r}
run_pedsim(chromosome = c("Chr1", "Chr02", "Chr03"), n.marker = c(50,50,50), 
           tot.size.cm = c(100,100,100), centromere = c(50, 50, 50),
           n.ind = 200, mk.types = c("A.H.B"), 
           n.types = c(150), pop = "F2", path.pedsim = "/home/cristiane/Programs/PedigreeSim/",
           name.mapfile = "mapfile.map", name.founderfile="founderfile.gen",
           name.chromfile="sim.chrom", name.parfile="sim.par",
           name.out="sim_cod_F2")
```

```{r}
pedsim2raw(cross="f2 intercross", genofile = "sim_cod_F2_genotypes.dat", 
           parent1 = "P1", parent2 = "P2", f1 = "F1", 
           out.file = "sim_cod_F2.raw", miss.perc = 0)
```

## New approach

```{r}
dataset <- read_onemap("sim_cod_F2.raw")

twopts <- rf_2pts(dataset)
seq1 <- make_seq(twopts, "all")
lod_sug <- suggest_lod(dataset)
lgs <- group(seq1, LOD= lod_sug) # Groups is correct

lg1 <- make_seq(lgs,1)
# Distance estimation
time <- system.time(map1 <- map(lg1)) # time spent 25.781 sec
save.obj <- list(map1, time)
save(save.obj, file = "save.obj4.RData")
```

```{r}
load("save.obj4.RData")
save.obj[[1]]

# Time spent
save.obj[[2]]
```

```{r}
# Ordering
ug1 <- ug(lg1)
rcd1 <- rcd(lg1)
seriation1 <- seriation(lg1)
record1 <- record(lg1)
mds1 <- mds_onemap(lg1)
order1 <- order_seq(lg1)
order1 <- make_seq(order1, "force")

p_ug_cod1 <- rf_graph_table(ug1)
p_rcd_cod1 <- rf_graph_table(rcd1)
p_ser_cod1 <- rf_graph_table(seriation1)
p_rec_cod1 <- rf_graph_table(record1)
p_mds_cod1 <- rf_graph_table(mds1)
p_map_cod1 <- rf_graph_table(map1)
p_order_cod1 <- rf_graph_table(order1)
```

## Old approach

```{r}
dataset <- read_onemap(inputfile = "sim_cod_F2.raw")

twopts <- rf_2pts(dataset)
seq1 <- make_seq(twopts, "all")
lod_sug <- suggest_lod(dataset)
lgs <- group(seq1, LOD= lod_sug) # Group is correct

lg1 <- make_seq(lgs,1)
# Distance estimation
time <- system.time(map1 <- map(lg1)) # time 0.015 sec

save.obj <- list(map1, time)
save(save.obj, file = "save.obj3.RData")
```

```{r}
load("save.obj3.RData")
save.obj[[1]]

# Time spent
save.obj[[2]]
```

```{r}
# Ordering
ug1 <- ug(lg1)
rcd1 <- rcd(lg1)
seriation1 <- seriation(lg1)
record1 <- record(lg1)
mds1 <- mds_onemap(lg1)
order1 <- order_seq(lg1)
order1 <- make_seq(order1, "force")

p_ug_cod_old1 <- rf_graph_table(ug1)
p_rcd_cod_old1 <- rf_graph_table(rcd1)
p_ser_cod_old1 <- rf_graph_table(seriation1)
p_rec_cod_old1 <- rf_graph_table(record1)
p_mds_cod_old1 <- rf_graph_table(mds1)
p_map_cod_old1 <- rf_graph_table(map1)
p_order_cod_old1 <- rf_graph_table(order1)
```

# Scenario 2

## Simulation

```{r}
run_pedsim(chromosome = c("Chr1", "Chr02", "Chr03"), n.marker = c(50,50,50), 
           tot.size.cm = c(100,100,100), centromere = c(50, 50, 50),
           n.ind = 200, mk.types = c("A.H.B", "C.A", "D.B"), 
           n.types = c(50, 50, 50), pop = "F2", 
           path.pedsim = "/home/cristiane/Programs/PedigreeSim/",
           name.mapfile = "mapfile.map", name.founderfile="founderfile.gen",
           name.chromfile="sim.chrom", name.parfile="sim.par",
           name.out="sim_F2")
```

```{r}
pedsim2raw(cross="f2 intercross", genofile = "sim_F2_genotypes.dat", 
           parent1 = "P1", parent2 = "P2", f1 = "F1", 
           out.file = "sim_F2.raw", miss.perc = 0)
```

## New approach

```{r}
dataset <- read_onemap(inputfile = "sim_F2.raw")

twopts <- rf_2pts(dataset)
seq1 <- make_seq(twopts, "all")
lod_sug <- suggest_lod(dataset)
lgs <- group(seq1, LOD= lod_sug) # Do not group correctly

lg1 <- make_seq(twopts,1:50)
# Distance estimation
time <- system.time(map1 <- map(lg1)) # Time spent 125.072 sec or 2.08 min
save.obj <- list(map1, time)

save(save.obj, file = "save.obj2.RData")
```

```{r}
load("save.obj2.RData")
save.obj[[1]]

# Time spent
save.obj[[2]]
```

```{r}
# Ordering
ug1 <- ug(lg1)
rcd1 <- rcd(lg1)
#seriation1 <- seriation(lg1) Error: There are too many ties in the ordering process - please, consider using another ordering algorithm.
record1 <- record(lg1)
mds1 <- mds_onemap(lg1)
order1 <- order_seq(lg1)
order1 <- make_seq(order1, "force")

p_ug1 <- rf_graph_table(ug1)
p_rcd1 <- rf_graph_table(rcd1)
#p_ser1 <- rf_graph_table(seriation1)
p_rec1 <- rf_graph_table(record1)
p_mds1 <- rf_graph_table(mds1)
p_map1 <- rf_graph_table(map1)
p_order1 <- rf_graph_table(order1)
```

## Old approach

```{r}
dataset <- read_onemap(inputfile = "sim_F2.raw")

twopts <- rf_2pts(dataset)
seq1 <- make_seq(twopts, "all")
lod_sug <- suggest_lod(dataset)
lgs <- group(seq1, LOD= lod_sug) # Do not group correctly

lg1 <- make_seq(twopts,1:50)
# Distance estimation
time <- system.time(map1 <- map(lg1)) # Time spent 2.150 sec

save.obj <- list(map1, time)
save(save.obj, file = "save.obj1.RData")
```

```{r}
load("save.obj1.RData")
save.obj[[1]]

# Time spent
save.obj[[2]]
```

```{r}
# Ordering
ug1 <- ug(lg1)
rcd1 <- rcd(lg1)
#seriation1 <- seriation(lg1) Error: There are too many ties in the ordering process - please, consider using another ordering algorithm.
record1 <- record(lg1)
mds1 <- mds_onemap(lg1)
order1 <- order_seq(lg1)
order1 <- make_seq(order1, "force")

p_ug_old1 <- rf_graph_table(ug1)
p_rcd_old1 <- rf_graph_table(rcd1)
#p_ser_old1 <- rf_graph_table(seriation1)
p_rec_old1 <- rf_graph_table(record1)
p_mds_old1 <- rf_graph_table(mds1)
p_map_old1 <- rf_graph_table(map1)
p_order_old1 <- rf_graph_table(order1)

save.image(file = "sim_old.RData")
```

# Ordering comparision


```{r, message=FALSE, warning=FALSE}
library(ggpubr)
load("sim_old.RData")

p <- ggarrange(p_ug_cod1, p_ug_cod_old1, p_ug1, p_ug_old1 , common.legend = TRUE,
          labels = c("ug sce1 new", "ug sce1 old", "ug sce2 new", "ug sce2 old"), 
          vjust = 0.3,
          hjust= -1.5,
          ncol=2, nrow=2)
p

p <- ggarrange(p_rcd_cod1, p_rcd_cod_old1, p_rcd1, p_rcd_old1 , common.legend = TRUE,
          labels = c("rcd sce1 new", "rcd sce1 old", "rcd sce2 new", "rcd sce2 old"), 
          vjust = 0.3,
          hjust= -1.5,
          ncol=2, nrow=2)
p

p <- ggarrange(p_rec_cod1, p_rec_cod_old1, p_rec1, p_rec_old1 , common.legend = TRUE,
          labels = c("rec sce1 new", "rec sce1 old", "rec sce2 new", "rec sce2 old"), 
          vjust = 0.3,
          hjust= -1.5,
          ncol=2, nrow=2)
p

p <- ggarrange(p_order_cod1, p_order_cod_old1, p_order1, p_order_old1 , common.legend = TRUE,
          labels = c("order sce1 new", "order sce1 old", "order sce2 new", "order sce2 old"), 
          vjust = 0.3,
          hjust= -1.5,
          ncol=2, nrow=2)
p

p <- ggarrange(p_mds_cod1, p_mds_cod_old1, p_mds1, p_mds_old1 , common.legend = TRUE,
          labels = c("mds sce1 new", "mds sce1 old", "mds sce2 new", "mds sce2 old"), 
          vjust = 0.3,
          hjust= -1.5,
          ncol=2, nrow=2)
p

```

