---
title: "SimulationsReads workflow"
date: "`r Sys.Date()`"
author: "[Statistical Genetics Lab](http://statgen.esalq.usp.br) <br/> Department of Genetics <br/> Luiz de Queiroz College of Agriculture <br/> University of SÃ£o Paulo"
output:
    rmdformats::readthedown:
      css: readthedownstatgen.css
---

Here will be describe how to use the onemap workflows. There are two main workflows:

* SimulatedReads.wdl

This one performs the reads simulations from a input chromosome sequence for the parents and progeny of a bi-parental population, the aligment of the reads, the SNP calling in gatk and freebayes softwares and genotype calling in updog, supermassa and polyrad, and linkage maps in onemap and gusmap.

![](images/SimulatedReads.svg)

* Simulate reads

input: reference genome
output: fastq files

![](images/reads_simu.svg)


* Create_aligments_from_families_files

input: fastq files
output: bam files

![](images/create_alignment_from_read_simulations.svg)


* freebayes_genotyping.wdl

input: bam files
output: vcf file

![](images/freebayes_genotyping.svg)

* gatk_genotyping.wdl

input: bam files
output: vcf file

![](images/gatk_genotyping.svg)

* default_maps.wdl

input: vcf files
output: RData and reports

![](images/default_maps.svg)

* SNPCaller_maps.wdl

input: vcf files
output: RData and reports

![](images/snpcaller_maps.svg)

* polyrad_maps.wdl

input: vcf files
output: RData and reports

![](images/polyrad_maps.svg)

* supermassa_maps.wdl

input: vcf files
output: RData and reports

![](images/supermassa_maps.svg)

* updog_maps.wdl

input: vcf files
output: RData and reports

![](images/updog_maps.svg)

* gusmap_maps.wdl

input: vcf files
output: RData and reports

![](images/gusmap_maps.svg)

You can also build your own workflow using tasks available in utils.wdl and utilsR.wdl
